metricbeat/module/zookeeper/server/server.go

```diff
@@ -0,0 +1,249 @@
+// specific language governing permissions and limitations
+// under the License.
+
+/*
+Package server fetches metrics from ZooKeeper by using the srvr command
+
+See the srvr command documentation at
+https://zookeeper.apache.org/doc/current/zookeeperAdmin.html
+
+ZooKeeper srvr Command Output
+
+  $ echo srvr | nc localhost 2181
+	Zookeeper version: 3.4.13-2d71af4dbe22557fda74f9a9b4309b15a7487f03, built on 06/29/2018 04:05 GMT
+Latency min/avg/max: 1/2/3
+Received: 46
+Sent: 45
+Connections: 1
+Outstanding: 0
+Zxid: 0x700601132
+Mode: standalone
+Node count: 4
+Proposal sizes last/min/max: -3/-999/-1
+
+
+*/
+package server
+
+import (
+	"bufio"
+	"encoding/binary"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/metricbeat/mb"
+	"github.com/elastic/beats/metricbeat/mb/parse"
+	"github.com/elastic/beats/metricbeat/module/zookeeper"
+	"github.com/pkg/errors"
+	"io"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+var latencyCapturer = regexp.MustCompile(`(\d+)/(\d+)/(\d+)`)
+var ipCapturer = regexp.MustCompile(`\d+\.\d+\.\d+\.\d+`)
+var thatNumberCapturer = regexp.MustCompile(`\[(\d+)\]`)
+var portCapturer = regexp.MustCompile(`:(\d+)\[`)
+var dataCapturer = regexp.MustCompile(`(\w+)=(\d+)`)
+var fieldsCapturer = regexp.MustCompile(`^([a-zA-Z\s]+):\s(\d+)`)
+var versionCapturer = regexp.MustCompile(`:\s(.*),`)
+var dateCapturer = regexp.MustCompile(`built on (.*)`)
+
+func init() {
+	mb.Registry.MustAddMetricSet("zookeeper", "server", New,
+		mb.WithHostParser(parse.PassThruHostParser),
+		mb.DefaultMetricSet(),
+	)
+}
+
+// MetricSet for fetching ZooKeeper health metrics.
+type MetricSet struct {
+	mb.BaseMetricSet
+}
+
+// New creates new instance of MetricSet.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	return &MetricSet{
+		BaseMetricSet: base,
+	}, nil
+}
+
+// Fetch fetches metrics from ZooKeeper by making a tcp connection to the
+// command port and sending the "srvr" command and parsing the output.
+func (m *MetricSet) Fetch(reporter mb.ReporterV2) {
+	outputReader, err := zookeeper.RunCommand("srvr", m.Host(), m.Module().Config().Timeout)
+	if err != nil {
+		reporter.Error(errors.Wrap(err, "srvr command failed"))
+		return
+	}
+
+	metricsetFields, err := parseSrvr(outputReader)
+	if err != nil {
+		reporter.Error(err)
+		return
+	}
+
+	reporter.Event(mb.Event{
+		MetricSetFields: metricsetFields,
+	})
+}
+
+func parseSrvr(i io.Reader) (common.MapStr, error) {
+	scanner := bufio.NewScanner(i)
+
+	output := common.MapStr{}
+
+	//Get version
+	ok := scanner.Scan()
+	if !ok {
+		return nil, errors.New("no initial successful scan, aborting")
+	}
+	output.Put("version.id", versionCapturer.FindStringSubmatch(scanner.Text())[1])
+	output.Put("version.date", dateCapturer.FindStringSubmatch(scanner.Text())[1])
+
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		if strings.Contains(line, "Zxid") {
+			xid, err := parseXid(line)
+			if err != nil {
+				return nil, errors.Wrap(err, "error parsing xid line")
```

Comment:
General question: If we have an error parsing a line, should we continue with the next line or return an error.

I'm leaning in the direction of continuing and logging on the debug level. Reason is if there is for a example a version where the parsing of one line does not work, metricbeat would completely stop working instead of just skipping that entry. For the error log it would be nice to include the actual line that could not be processed for debugging purpose.

Reply:
Yes, I also thing that very descriptive errors and detailed handling is a _must_. I just saw that it's not the case in some modules and I thought it was a general rule to avoid filling the logs with error messages.

I have updated the error messages and now partial content could be pushed by omitting lines with errors. :+1: